### PR TITLE
CI: macOS3.9 testing

### PR DESCRIPTION
**Pull request by andyfaff**

### CODE REVIEW

**Comment by tylerjereddy** [comment]
Ah, simple availability issue it seems--hopefully resolved soon upstream

**Comment by tylerjereddy** [comment]
Marking for eventual backport as 3.9 + MacOS will be useful on the 1.5.x branch as well.

**Comment by tylerjereddy** [comment]
Restarting the GitHub actions MacOS jobs to see if Python 3.9 is working there now--things are starting to move into production for the major services with 3.9 now so fingers crossed.

**Comment by tylerjereddy** [comment]
Python 3.9 is available in there--I believe the issue now is that without pre-built NumPy 3.9 wheels NumPy gets built with whatever linear algebra backend is in the CI environment and this is causing an issue. I believe Chuck is moving forward with wheels for 3.9 soon; so, we'll probably follow a bit after that. Error below:

```
Collecting scikit-umfpack
  Downloading scikit-umfpack-0.3.2.tar.gz (25 kB)
    ERROR: Command errored out with exit status 1:
     command: /Users/runner/hostedtoolcache/Python/3.9.0/x64/bin/python -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/private/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/pip-install-frmv2eke/scikit-umfpack/setup.py'"'"'; __file__='"'"'/private/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/pip-install-frmv2eke/scikit-umfpack/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' egg_info --egg-base /private/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/pip-pip-egg-info-xjpxg_yp
         cwd: /private/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/pip-install-frmv2eke/scikit-umfpack/
    Complete output (14 lines):
    /private/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/pip-install-frmv2eke/scikit-umfpack/setup.py:56: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
      import imp
    init_dgelsd failed init
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "/private/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/pip-install-frmv2eke/scikit-umfpack/setup.py", line 250, in <module>
        setup_package()
      File "/private/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/pip-install-frmv2eke/scikit-umfpack/setup.py", line 188, in setup_package
        import numpy
      File "/Users/runner/hostedtoolcache/Python/3.9.0/x64/lib/python3.9/site-packages/numpy/__init__.py", line 286, in <module>
        raise RuntimeError(msg)
    RuntimeError: Polyfit sanity test emitted a warning, most likely due to using a buggy Accelerate backend. If you compiled yourself, see site.cfg.example for information. Otherwise report this to the vendor that provided NumPy.
    RankWarning: Polyfit may be poorly conditioned
```

**Comment by tylerjereddy** [comment]
NumPy 3.9 wheels released in the last hour or so--re-trying MacOS 3.9 in CI again

**Comment by andyfaff** [comment]
@tylerjereddy, gets a bit further, but now there's a Fortran compile error

**Comment by andyfaff** [comment]
I think the fixes required here those that #12957 are currently failing on.

**Comment by tylerjereddy** [comment]
I see, let me work on this a bit--there was some indication that the issue was fixed on `master` but this is suggesting otherwise I guess.

**Comment by tylerjereddy** [comment]
I'll try applying that patch here to see if it helps; locally I can build this feature branch and `master` with some version of XCode 12 I think, but let's see what happens here..

**Comment by tylerjereddy** [comment]
some of the weird changes in gh-12243 may be involved? @sethtroisi ?

I probably won't solve this tonight..

**Comment by sethtroisi** [comment]
I re-reviewed #12243 and didn't see anything weird in that PR. I looked at the errors here and they seem in unrelated files. I suspect that a compiler or default warnings changed?

**Comment by tylerjereddy** [comment]
@sethtroisi 

The word "weird" is used to describe a change related to one of the errors here:

> slacon2.c: idamax_ changed to isamax_

And the error in CI here:

```
   scipy/sparse/linalg/dsolve/SuperLU/SRC/slacon2.c:160:16: error: implicit declaration of function 'idamax_' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
      isave[1] = idamax_(n, &x[0], &c__1);  /* j */
```

Those are really completely independent things even though it is the same file and the same variable involved?

**Comment by sethtroisi** [comment]
@tylerjereddy Sorry, I didn't see that error before. My change is certainly suspicious. It's a little weird because if you trace it to the upstream repository (https://github.com/xiaoyeli/superlu/pull/19) It was changed it from `isamax_` because of `-Wimplicit-function-declaration`. It's possible superlu is not tested on mac and `idamax_` is `#ifdef` behind some linux flag. 

---

Looking at one of the other errors
```
  scipy/sparse/linalg/dsolve/SuperLU/SRC/dmach.c:62:9: error: implicitly declaring library function 'strncmp' with type 'int (const char *, const char *, unsigned long)' [-Werror,-Wimplicit-function-declaration]
```
I'm surprised by this. https://github.com/scipy/scipy/pull/12243/files#diff-7976f5e4c73cb640a7a8c719dc525d911720b1c28965a8f55894f81835e0a5b5
 added `include <string.h>`  which I would have thought would fix this.

**Comment by tylerjereddy** [comment]
I've probably messed up the patch added here too--it was cherry-picked from a branch intended for 1.5.x instead of master. Anyway, I just need to be a bit cautious and if there's advice that's helpful. I'll try to sort out something soon. I wonder if the CI here is using yet another version of XCode or something.

**Comment by andyfaff** [comment]
I don't mind helping, but I'm not sure how helpful I can be. Would knowing what version of the compilers we are using be helpful?

**Comment by tylerjereddy** [comment]
>  Would knowing what version of the compilers we are using be helpful?

Probably! Just scanning through the raw log I see `XCODE_12_DEVELOPER_DIR=/Applications/Xcode_12.1.app/Contents/Developer`, so XCode 12.1 seems like a logical guess given we've seen this issue reported alongside bleeding edge XCode.

I have the same version locally, maybe I need to check the install command I use vs. CI for reproducing or something.

**Comment by tylerjereddy** [comment]
@andyfaff As noted in gh-12957, the failure here for MacOS Python 3.9 + XCode 12 is at the dependency install stage--some dependency is pulling in its own version of SciPy rather than the failure happening for the branch under test:

```
Collecting scipy>=1.0.0rc1
  Downloading scipy-1.5.3.tar.gz (25.2 MB)
  Installing build dependencies: started
  Installing build dependencies: still running...
  Installing build dependencies: still running...
  Installing build dependencies: finished with status 'done'
  Getting requirements to build wheel: started
  Getting requirements to build wheel: finished with status 'done'
```

Is it acceptable/normal for us to have a dependency that depends on us during i.e., its pip build? That might create a bit of a chicken and egg situation.. I guess we need to identify which dependency is doing that and adjust it.

**Comment by andyfaff** [comment]
That's weird. When I set this up I copied the dependencies over from Travis. The only dependencies that we would need beforehand are numpy, cython, pybind11, openblas and any build tools. Anything would be for testing purposes afterwards.

**Comment by andyfaff** [comment]
@tylerjereddy  scikit-umfpack needs scipy. I've tried removing that from the package install step.

**Comment by andyfaff** [comment]
It's a chicken and egg situation. scikit-umfpack wants to install scipy as a dependent package, but there isn't a wheel available to install, so it tries to build it from 1.5.3 source. We use it to test various scipy aspects for source>1.5.3. Tests are now running.

**Comment by andyfaff** [comment]
@tylerjereddy, yay, tests pass.

**Comment by tylerjereddy** [comment]
> I'll merge tomorrow, unless you have any objections.

Do you mind if we back out all of the shims I added for superlu and qhull and so on? I don't think `master` needs any of them based on local testing so if we can avoid modifying that vendored code seems good right? 

I'm definitely going to backport the CI addition, so if we could modify the PR to just have one commit with your changes to `.github/workflows/testing.yml` that would be ideal. Either of us could do that easily enough if we agree here.

Then when `1.5.4` is out we can try turning back on the dep(s) that depend on us.

File: .github/workflows/testing.yml [hunk 1]
```diff
@@ -86,7 +86,7 @@ jobs:
     - name: Install packages
       run: |
         pip install ${{ matrix.numpy-version }}
-        pip install setuptools wheel cython gmpy2 pytest pytest-xdist pybind11 pytest-xdist mpmath scikit-umfpack scikit-sparse
+        pip install setuptools wheel cython pytest pytest pytest-xdist pybind11 pytest-xdist mpmath gmpy2
```
Comment (tylerjereddy) [inline .github/workflows/testing.yml:89]: looks like `pytest` is showing up twice; either of us can fix that easily enough

**Comment by tylerjereddy** [comment]
If the CI is all green feel free to merge of course, then I'll proceed with the backport

**Pull request merged**